### PR TITLE
Fix Boolean schemas

### PR DESCRIPTION
**Pull request by deBhal**
This PR fixes a bug in `is-my-json-valid` where the boolean schema `false` is incorrectly treated as `true`.

According to the the [spec](http://json-schema.org/latest/json-schema-core.html#rfc.section.4.3.1), `false` is a valid schema that will not match any data, but `is-my-json-valid` incorrectly treats it as an empty schema, meaning it accepts anything instead:

```js
var validate = require('is-my-json-valid')(false);
[null, 12, "a string", {foo: "bar"}].map(validate);
// => [ true, true, true, true ]
```

With no explicit handling for boolean schemas, and 
Since true and false are objects in javascript, having no explicit handling means they're both treated like `{}`, which is a happy coincidence for `true`, and a less fortunate one for `false` :)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -171,6 +171,10 @@ var compile = function(schema, cache, root, reporter, opts) {
       }
     }
 
+    if (node === false) {
```
Comment (sirreal) [inline index.js:174]: In addition to failing early with `false` schemas, we could potentially save a significant amount of work by passing early for `true` schemas. I'm not familiar enough with the codebase to know whether that could be significant or not, but keep it in mind.

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "is-my-json-valid",
-  "version": "2.16.1",
+  "version": "2.17.1",
```
Comment (sirreal) [inline package.json]: A version change for the `false` and `true` schemas should probably be a breaking `v3` change. It will break any schemas relying on the bug.

File: index.js (hunk 1 above)
Comment (deBhal) [reply index.js:174]: It's a great point, I'm just scared of falling into the performance rabbit hole. I already started slipping, and wrote https://github.com/deBhal/is-my-json-valid/pull/2 by way of backstop :) 

Let's continue this discussion there.

**Comment by deBhal** [comment]
Unfortunately, the link I was looking at - http://json-schema.org/latest/json-schema-core.html#rfc.section.4.3.1 is draft-07, and `is-my-json-valid` validates against [draft-04](https://tools.ietf.org/html/draft-wright-json-schema-00#section-4.4), which says:

>  A JSON schema MUST be an object.

Boolean schemas were added in draft-06 (see [additions-and-backwards-compatible-changes])(http://json-schema.org/draft-06/json-schema-migration-faq.html#additions-and-backwards-compatible-changes)).

This recasts this change from spec compliance bugfix to a partial upgrade to a more recent spec, which is a much less straightforward proposition.

This specific issue is from the "additions-and-backwards-compatible-changes" ([#](http://json-schema.org/draft-06/json-schema-migration-faq.html#additions-and-backwards-compatible-changes)) section, but it seems like a breaking change to me - any schema that includes `false` as a schema is going to behave differently according to the different schemas.

I had a look to see if upgrading to draft7 properly is an option (#3), and I think we'd want some pretty compelling use cases before we take that on.

TL;DR: I don't see this change landing upstream. How badly do we need this functionality?

**Comment by sirreal** [comment]
This is a tricky situation.

> > A JSON schema MUST be an object.

I found a reference from [draft-04 here](https://tools.ietf.org/html/draft-zyp-json-schema-04#section-3.2):

> 3.2.  JSON Schema, keywords
>
>   A JSON Schema is a JSON document, and that document MUST be an
   object.  Object members (or properties) defined by JSON Schema (this
   specification, or related specifications) are called keywords, or
   schema keywords.

Does `is-my-json-valid` comply with this now? I don't think so, it behaves rather strangely. I whipped up a [gist with some tests](https://gist.github.com/sirreal/a57c2901a004053488b212ea6ca3510e) and it does not behave as expected for several cases:

```
  is-my-json-valid schema must be object
    ✅  object 👍 (1ms)
    ✅  Object 👍 (1ms)
    ✅  undefined 👎
    ❗️  array 👎 (1ms)
    ✅  null 👎
    ❗️  boolean 👎
    ✅  string 👎 (1ms)
    ❗️  number 👎 (1ms)
    ❗️  regexp 👎
    ❗️  Boolean 👎
    ❗️  Number 👎 (1ms)
    ❗️  String 👎
    ❗️  RegExp 👎
    ❗️  Array 👎 (1ms)
    ❗️  Set 👎
    ❗️  Map 👎
```

Unfortunately, I'm not entirely sure what the spec means by `object` here. Strictly an `Object` as in `object !== null && typeof object === 'object' && object.constructor === Object;`? Or a definition more like "has property access"?

Right now, `is-my-json-valid` seems to use the latter approach. The most surprising thing for me is that number and regexp literals, i.e.`1` or `/.*/`, are parsed as valid schemas. I wouldn't expect `1` to have property access 😕 

> This specific issue is from the "additions-and-backwards-compatible-changes" section, but it seems like a breaking change to me - any schema that includes `false` as a schema is going to behave differently according to the different schemas.

My interpretation of this is that `is-my-json-valid` does not strictly comply with draft-04, where boolean schemas would not be valid. If boolean schemas were invalid, then I understand that adding support for them would be a a backwards compatible addition.

> TL;DR: I don't see this change landing upstream. How badly do we need this functionality?

We don't have a use case for this right now. The boolean schemas that have been introduced were actually developer mistakes. However, they parsed without error and passed all inputs as valid. To be honest, it would be more useful if these threw errors on parse.

**Comment by deBhal** [comment]
> Does is-my-json-valid comply with this now?

I was interpreting this through a lens coloured by C & C++, where if you fail to meet the preconditions the behaviour is "undefined". So I would say it does, technically, comply but in the least helpful way possible :p

> We don't have a use case for this right now. The boolean schemas that have been introduced were actually developer mistakes.

Oh good, that's a bit of a relief. We could do this, but none of the options are very appealing. I'm going to go ahead and close this.

However, they parsed without error and passed all inputs as valid. To be honest, it would be more useful if these threw errors on parse.

> To be honest, it would be more useful if these threw errors on parse.

That seems pretty doable; The schema spec is also a schema. I'll have a quick look into it.

Note to self: Just in case we do come back to this, this implementation is not sufficient for some more complex cases. `{ "properties": {"foo": false }}` incorrectly fails when "foo" is missing, and there were also some boolean schema failures from the draft-7 suite that looked legitimate.

**Comment by deBhal** [comment]
> That seems pretty doable; The schema spec is also a schema. I'll have a quick look into it.

The validation is pretty easy - https://github.com/Automattic/wp-calypso/pull/21125. The bundling seems trickier. I could figure it out eventually, but maybe we can find someone who's more familiar with it.
